### PR TITLE
add duplicated_keys

### LINKED ISSUE #2004: Warn for duplicate keys in map literal
When I write
```dart
final uvIndexes = {
  1: 'low',
  1: 'moderate',
  3: 'high',
};
```

I would like a lint that warns me that I have a duplicate key in my map because it is most likely a bug.
The code is legal but chances are very low that is this intentional.

Note: it does warn already for `const` map.

### PR DESCRIPTION
**Pull request by a14n**
# Description

Warn for duplicate keys in set/map literals

Fixes #2004

### CODE REVIEW

File: lib/src/rules/duplicated_keys.dart [hunk 1]
```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2020, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import '../analyzer.dart';
+
+const _desc = r"Two keys in a constant set/map shouldn't be equal.";
```
Comment (bwilkerson) [inline lib/src/rules/duplicated_keys.dart]: The description says "constant", but neither the examples nor the tests use `const`. So I suspect that you mean "Two constant-valued elements / keys in a set / map literal shouldn't be equal.".

File: lib/src/rules/duplicated_keys.dart [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  void visitSetOrMapLiteral(SetOrMapLiteral node) {
+    final expressions = node.isSet
+        ? node.elements.whereType<Expression>()
+        : node.elements.whereType<MapLiteralEntry>().map((e) => e.key);
+    final alreadySeen = [];
+    for (final expression in expressions) {
+      final constEvaluation = context.evaluateConstant(expression);
+      if (constEvaluation.errors.isNotEmpty) continue;
+      final dartObject = constEvaluation.value;
+      if (alreadySeen.contains(dartObject)) {
```
Comment (bwilkerson) [inline lib/src/rules/duplicated_keys.dart]: If you made `alreadySeen` a set, then (a) lookup would probably be faster with large literals and (b) you could use `add` here which would also be more efficient.

**Comment by pq** [comment]
I'm w/ @bwilkerson: this feels like a great candidate for a hint.

Thanks, as always, @a14n!

**Comment by bwilkerson** [comment]
Unless you have an argument for why it shouldn't be a hint, we'd like to make it a hint. If we do, do you want to do that work, or do you want one of us to do that?

**Comment by a14n** [comment]
Please do it!

**Pull request closed without merging**